engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/UpdateByWindowTicks.java

```diff
@@ -0,0 +1,418 @@
+                for (int ii = 0; ii < chunkRsSize; ii++) {
+                    // read the current position
+                    final long currentPos = posChunk.get(ii);
+
+                    // compute the head and tail positions (inclusive)
+                    final long head = Math.max(0, currentPos - prevUnits + 1);
+                    final long tail = Math.min(sourceRowSetSize - 1, currentPos + fwdUnits);
+
+                    // pop out all values from the current window that are not in the new window
+                    int popCount = 0;
+                    while (!ctx.currentWindowPositions.isEmpty() && ctx.currentWindowPositions.front() < head) {
+                        ctx.currentWindowPositions.remove();
+                        popCount++;
+                    }
+
+                    // skip values until they match the window (this can only happen on initial addition of rows
+                    // to the table, because we short-circuited the precise building of the influencer rows for
+                    // efficiency)
+                    while (ctx.nextInfluencerPos < head) {
+                        ctx.nextInfluencerIndex++;
+
+                        if (ctx.nextInfluencerIndex < ctx.influencerPosChunkSize) {
+                            ctx.nextInfluencerPos = (int) ctx.influencerPosChunk.get(ctx.nextInfluencerIndex);
+                            ctx.nextInfluencerKey = ctx.influencerKeyChunk.get(ctx.nextInfluencerIndex);
+                        } else {
+                            // try to bring in new data
+                            loadNextInfluencerChunks(ctx);
+                        }
+                    }
+
+                    // push matching values
+                    int pushCount = 0;
+                    while (ctx.nextInfluencerPos <= tail) {
+                        ctx.currentWindowPositions.add(ctx.nextInfluencerPos);
+                        pushCount++;
+                        // add this key to the needed set for this chunk
+                        chunkInfluencerBuilder.appendKey(ctx.nextInfluencerKey);
+                        ctx.nextInfluencerIndex++;
+
+                        if (ctx.nextInfluencerIndex < ctx.influencerPosChunkSize) {
+                            ctx.nextInfluencerPos = (int) ctx.influencerPosChunk.get(ctx.nextInfluencerIndex);
+                            ctx.nextInfluencerKey = ctx.influencerKeyChunk.get(ctx.nextInfluencerIndex);
+                        } else {
+                            // try to bring in new data
+                            loadNextInfluencerChunks(ctx);
+                        }
+                    }
+
+                    // write the push and pop counts to the chunks
+                    popChunk.set(ii, popCount);
+                    pushChunk.set(ii, pushCount);
+                }
+
+                // execute the operators
+                try (final RowSet chunkInfluencerRs = chunkInfluencerBuilder.build()) {
```

Comment:
I think you can greatly simplify by not reading chunks of influencer keys and building from them, but rather instead figuring out how many you need and getting that many from the iterator over influencer keys.